addons/functions/functions/fnc_bpzBuild.sqf

```diff
@@ -17,57 +17,77 @@
  * Public: No
  */
 
-params ["_vehicle", "_trenchClass", "_offset"];
+params ["_vehicle"];
 
-private _trench = _trenchClass createVehicle [0,0,0];
-_trench animateSource ["rise", 0, true];
-_trench attachTo [_vehicle, _offset];
-_trench setObjectTexture [0, surfaceTexture getPos _vehicle];
-_trench setVariable ["ace_trenches_progress", 0, true];
-
-private _digTime = missionNamespace getVariable [getText (configFile >> "CfgVehicles" >> _trenchClass >>"ace_trenches_diggingDuration"), 20];
+private _config = configFile >> "CfgDigVehicles" >> typeOf _vehicle;
+private _distanceToTrench = getNumber (_config >> "distanceToTrench");
 
 [{
     params ["_args", "_handle"];
-    _args params ["_vehicle", "_trench", "_digTime"];
-
-    private _actualProgress = _trench getVariable ["ace_trenches_progress", 0];
+    _args params ["_vehicle", "_distanceToTrench"];
 
     if (
         isNull _vehicle ||
-        !(alive _vehicle) ||
-        _actualProgress >= 1
+        !(alive _vehicle)
     ) exitWith {
         [_handle] call CBA_fnc_removePerFrameHandler;
-        detach _trench;
-        _trench setVariable ["ace_trenches_placeData", [getPos _trench, [vectorDir _trench, vectorUp _trench]], true];
     };
 
-    if (_actualProgress <= 0) exitWith {
-        [_handle] call CBA_fnc_removePerFrameHandler;
-        deleteVehicle _trench;
-    };
+    // save performance
+    if (!(_vehicle getVariable ["grad_trenches_functions_plowlowered", false]) && !(_vehicle getVariable [QGVAR(isDigging), false])) exitWith {};
 
-    private _speedLimit = 15;
-    private _speed = speed _vehicle;
+    // init digging
+    if (!(_vehicle getVariable [QGVAR(isDigging), false])) then {
 
-    if (_speed > _speedLimit) then {
-      _vehicle setVelocity ((velocity _vehicle) vectorMultiply ((_speedLimit / _speed) - 0.00001));
-    };
+        // can dig vehicle on this position
+        if (!([_vehicle modelToWorld [-0.227,_distanceToTrench,0]] call FUNC(canDig))) exitWith {};
+
+        // only work when vehicle is not tilted
+        if (vectorUp _vehicle select 2 < 0.99) exitWith {};
 
-    private _animationPhase = _trench animationSourcePhase "rise";
-    private _diff = _trench getVariable [QGVAR(diggingSteps), ([configOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry) / (_digTime * 10)];
+        private _speed = speed _vehicle;
+        if (_speed > 1) then {   
 
-    if (_speed > 1) then {
-        _trench animateSource ["rise", _animationPhase + _diff, true];
+            _vehicle setVariable [QGVAR(isDigging), true, true];
+            private _trench = "GRAD_envelope_vehicle" createVehicle [0,0,0];
```

Comment:
Why only limit it to this one trench type?

Reply:
ease of use (you would need a submenu to select a trench type which you can not really see as a driver anyway, so you easily misconfigure), looks (trench should not be much wider than vehicle or way smaller) and reasonability (if you dig with a vehicle the result will look the same mostly, as you dont have much control for small adjustments)

Reply:
@AdlerSalbei  would you be okay with merging in this state and check for user feedback / needs? (successful dedicated test provided)

Reply:
We can merge it as is, but i'll properly extend it in the future.